### ISSUE TITLE
parameterize reactor in twisted.application.internet services

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2937]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2937|
|Type|enhancement|
|Created|2007-12-06 14:08:10Z|
|Branch|https://github.com/twisted/twisted/tree/service-reactor-2937|

The unpleasant way ports are bound and connections are made by the internet services means that it is difficult or impossible to test this code without actually hitting the network.  This is likewise for application code (such as ```makeService``` implementations) which creates these objects.

It should be possible to specify the reactor used by these classes.

<details><summary>Searchable metadata</summary>

```
trac-id__2937 2937
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_service_reactor_2937 branches/service-reactor-2937
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1196950090000000 1196950090000000
changetime__1198845074000000 1198845074000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [22037]) Branching to 'service-reactor-2937'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22038]) Make reactor parametrizable in clients and servers.

Refs [#2937](https://github.com/twisted/twisted/issues/2937)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

This is ready to review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

The implementations of _getPort and _getConnection won't work as expected the second time they're invoked (eg, startService, stopService, startService).

I guess there should be some test coverage for the repeated use of the same service object.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [22051]) Remove the 'reactor' keyword in init, add tests for multiple start, some docs.

Refs [#2937](https://github.com/twisted/twisted/issues/2937)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Oups, thanks, I had missed that. Back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

pyflakes warnings on ```test_application.py```:

  * twisted/test/test_application.py:20: 'microdom' imported but unused
  * twisted/test/test_application.py:591: redefinition of fuction 'test_reactorParametrizationInClient' from line 553

```_getPort``` and ```_getConnection``` could probably use some docstrings.

Other than those issues, I think it looks good.  No need for re-review unless you notice something else.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [22161]) Rename a test, add some docstrings to internal methods.

Refs [#2937](https://github.com/twisted/twisted/issues/2937)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [22163]) Merge service-reactor-2937

Author: therve
Reviewer: exarkun
Fixes [#2937](https://github.com/twisted/twisted/issues/2937)

Add a reactor keyword for services in twisted.application.internet
so that we can test this code more easily without hitting the network.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
